README.md
forge
=====

[![Build Status](https://travis-ci.org/go-aah/forge.svg?branch=master)](https://travis-ci.org/go-aah/forge)
[![codecov](https://codecov.io/gh/go-aah/forge/branch/master/graph/badge.svg)](https://codecov.io/gh/go-aah/forge)
[![GoDoc](https://godoc.org/github.com/go-aah/forge?status.svg)](https://godoc.org/github.com/go-aah/forge)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)

Forge is a configuration syntax and parser forked for [aah framework](https://aahframework.org).

## Installation

```bash
require github.com/go-aah/forge v0.8.0
```

## Example

You can see example usage in the `example` folder.

```cfg
# example.cfg

# Global directives
global = "global value";

# Pull in the "PATH" environment variable
path_env = $PATH;

# Primary section
primary {
  string = "primary string value";
  single = 'single quotes are allowed too';

  # Semicolons are optional
  integer = 500
  float = 80.80
  boolean = true
  negative = FALSE
  nothing = NULL

  list = [50.5, true, false, "hello", 'world'];

  # Include external files
  include "./include*.cfg";
  # Primary-sub section
  sub {
      key = "primary sub key value";
  }
}

# Secondary section
secondary {
  another = "secondary another value";
  global_reference = global;
  primary_sub_key = primary.sub.key;
  another_again = .another;  # References secondary.another
  _under = 50;
}
```

```go
package main

import (
	"fmt"
	"json"

	"github.com/go-aah/forge"
)

func main() {
	// Parse a `SectionValue` from `example.cfg`
	settings, err := forge.ParseFile("example.cfg")
	if err != nil {
		panic(err)
	}

	// Get a single value
	if settings.Exists("global") {
		// Get `global` casted as a string
		value, _ := settings.GetString("global")
		fmt.Printf("global = \"%s\"\r\n", value)
	}

	// Get a nested value
	value, err := settings.Resolve("primary.included_setting")
	fmt.Printf("primary.included_setting = \"%s\"\r\n", value.GetValue())

	// You can also traverse down the sections manually
	primary, err := settings.GetSection("primary")
	strVal, err := primary.GetString("included_setting")
	fmt.Printf("primary.included_setting = \"%s\"\r\n", strVal)

	// Convert settings to a map
	settingsMap := settings.ToMap()
	fmt.Printf("global = \"%s\"\r\n", settingsMap["global"])

	// Convert settings to JSON
	jsonBytes, err := settings.ToJSON()
	fmt.Printf("\r\n\r\n%s\r\n", string(jsonBytes))
}
```

## Issues/Requests?

Please feel free to open a github issue for any issues you have or any feature requests.


CHANGELOG.md
**v0.8.0** 2018-12-12 Jeevanandam M <jeeva@myjeeva.com>
	* Handling external and native fs more effectively

**v0.7.0** 2018-12-07 Jeevanandam M <jeeva@myjeeva.com>
	* Removed vfs module dependency
	* Added FileSystem interface to extending possibility  
...

**v0.5** 2018-06-20 Jeevanandam M <jeeva@myjeeva.com>
	* Virtual FileSystem capability added `aahframework.org/vfs`

**v0.4.4** 2017-03-22 Jeevanandam M <jeeva@myjeeva.com>
	* Upstream Refresh
	* File closer enhanced
	* Relative subdirectory support added for `include`

**v0.4 to v0.4.3** 2017-03-02 Jeevanandam M <jeeva@myjeeva.com>
* Package versioning tested and released

**v0.3.2** 2017-03-02 Jeevanandam M <jeeva@myjeeva.com>
* Package name updated to aahframework.org/*
* Travis build config update
* Resolving relative path for `include` tag to parent conf directory
* Fix line no and EOF check for scanner

**v0.3.1** 2016-09-02 Jeevanandam M <jeeva@myjeeva.com> (Upstream refresh from v0.2.1)
* Add new `func (*Section) Merge` to merge multiple `Section`s together.

**v0.3.0**	2016-07-02	Jeevanandam M <jeeva@myjeeva.com>
* Forked `github.com/brettlangdon/forge` into `go-aah` org, please refer [here](https://github.com/brettlangdon/forge/issues/13#issuecomment-229080913) why?
* Merged PR [brettlangdon/forge#19](https://github.com/brettlangdon/forge/pull/19/files) from upstream and made required changes
* Version bumped for merging new feature environment variable support
* I'm continuing version number to honor his efforts made nice library `forge` which is similar to `HOCON` syntax but forge direction is different :)

-----

2016-02-26  Brett Langdon  <me@brett.is>

	* v0.2.0: Fix bug with parsing of lists without ending semicolon

	Note: We bumped to `v0.2.0` since we were not doing a great job of following
	semver in previous releases, this version changes aims to correct that mistake.

2015-10-02  Brett Langdon  <me@brett.is>

	* v0.1.7: Allow empty string values

2015-08-09  Brett Langdon  <me@brett.is>

	* v0.1.6: Allow integers in identifiers

2015-07-15  Brett Langdon  <me@brett.is>

	* v0.1.5: Add support for lists

2015-07-14  Brett Langdon  <me@brett.is>

	* v0.1.4: Make semicolons optional

2015-07-14  Brett Langdon  <me@brett.is>

	* v0.1.3: Allow single quoted strings

2015-07-04  Brett Langdon  <me@brett.is>

	* v0.1.2: Allow escaped double quotes and backslashes in strings

2015-06-28  Brett Langdon  <me@brett.is>

	* v0.1.1: Add `Section.Keys` method

2015-06-22  Brett Langdon  <me@brett.is>

	* v0.1.0: Initial release


forge.go
// Package forge provides an api to deal with parsing configuration files.
//
// Config file example:
//
//     # example.cfg
//     top_level = "a string";
//     primary {
//       primary_int = 500;
//       sub_section {
//         sub_float = 50.5;  # End of line comment
//       }
//     }
//     secondary {
//       secondary_bool = true;
//       secondary_null = null;
//
//       list = [
//         "any",
//         'value',
//         true,
//         55.5,
//       ]
//
//       # Reference other config value
//       local_ref = .secondary_null;
//       global_ref = primary.sub_section.sub_float;
//
//       # Include all files matching the provided pattern
//       include "/etc/app/*.cfg";
//     }
//
//
// Config file format:
//
//     IDENTIFIER: [_a-zA-Z]([_a-zA-Z0-9]+)?
//     NUMBERS: [0-9]+
//     END: ';' | '\n'
//
//     BOOL: 'true' | 'false'
//     NULL: 'null'
//     INTEGER: ('-')? NUMBERS
//     FLOAT: ('-')? NUMBERS '.' NUMBERS
//     STRING: ['"] .* ['"]
//     REFERENCE: (IDENTIFIER)? ('.' IDENTIFIER)+
//     ENVIRONMENT: '$' IDENTIFIER
//     VALUE: BOOL | NULL | INTEGER | FLOAT | STRING | REFERENCE | ENVIRONMENT
//     LIST: '[' (VALUE | LIST) (',' NEWLINE* (VALUE | LIST))+ ']'
//
//     INCLUDE: 'include ' STRING END
//     DIRECTIVE: (IDENTIFIER '=' (VALUE | LIST) | INCLUDE) END
//     SECTION: IDENTIFIER '{' (DIRECTIVE | SECTION)* '}'
//     COMMENT: '#' .* '\n'
//
//     CONFIG_FILE: (COMMENT | DIRECTIVE | SECTION)*
//
//
// Values
//  * String:
//      Any value enclosed in double or single quotes (e.g. "string" or 'string').
//      Double quotes, single quotes, and backslashes can be escaped with backslashes (e.g. "\"quoted\"", '\'quoted\'', and "\\<--backslash")
//  * Integer:
//      Any number without decimal places (e.g. 500)
//  * Float:
//      Any number with decimal places (e.g. 500.55)
//  * Boolean:
//      The identifiers 'true' or 'false' of any case (e.g. TRUE, True, true, FALSE, False, false)
//  * Null:
//      The identifier 'null' of any case (e.g. NULL, Null, null)
//  * List:
//      A list value is any number of other values separated by commas and surrounded by brackets.
//      (e.g. [50.5, 'some', "string", true, false])
//  * Global reference:
//      An identifier which may contain periods, the references are resolved from the global
//      section (e.g. global_value, section.sub_section.value)
//  * Local reference:
//      An identifier which main contain periods which starts with a period, the references
//      are resolved from the settings current section (e.g. .value, .sub_section.value)
//  * Environment:
//      An environment is a way to pull in environment variables into your config. Environment variables
//      are identifiers which start with a dollar sign (e.g. $PATH). Environment variables are always
//      represented as strings and are evaluated at parse time.
//
// Directives
//  * Comment:
//      A comment is a pound symbol ('#') followed by any text any which ends with a newline (e.g. '# I am a comment\n')
//      A comment can either be on a line of it's own or at the end of any line. Nothing can come after the comment
//      until after the newline.
//  * Directive:
//      A directive is a setting, a identifier and a value. They are in the format '<identifier> = <value>;'
//      All directives must end in either a semicolon or newline. The value can be any of the types defined above.
//  * Section:
//      A section is a grouping of directives under a common name. They are in the format '<section_name> { <directives> }'.
//      All sections must be wrapped in braces ('{', '}') and must all have a name. They do not end in a semicolon.
//      Sections may be left empty, they do not have to contain any directives.
//  * Include:
//      An include statement tells the config parser to include the contents of another config file where the include
//      statement is defined. Includes are in the format 'include "<pattern>";'. The <pattern> can be any glob
//      like pattern which is compatible with `path.filepath.Match` http://golang.org/pkg/path/filepath/#Match
//
package forge

import (
	"bytes"
	"io"
	"strings"
)

// Version represent current release version of forge.
var Version = "0.8.0"

// ParseBytes takes a []byte representation of the config file, parses it
// and responds with `*Section` and potentially an `error` if it cannot
// properly parse the config
func ParseBytes(data []byte) (*Section, error) {
	return ParseReader(bytes.NewReader(data))
}

// ParseFile takes a string filename for the config file, parses it
// and responds with `*Section` and potentially an `error` if it cannot
// properly parse the configf
func ParseFile(filename string) (*Section, error) {
	parser, err := NewFileParser(filename)
	if err != nil {
		return nil, err
	}
	err = parser.Parse()
	if err != nil {
		return nil, err
	}

	return parser.GetSettings(), nil
}

// ParseReader takes an `io.Reader` representation of the config file, parses it
// and responds with `*Section` and potentially an `error` if it cannot
// properly parse the config
func ParseReader(reader io.Reader) (*Section, error) {
	parser := NewParser(reader)
	err := parser.Parse()
	if err != nil {
		return nil, err
	}

	return parser.GetSettings(), nil
}

// ParseString takes a string representation of the config file, parses it
// and responds with `*Section` and potentially an `error` if it cannot
// properly parse the config
func ParseString(data string) (*Section, error) {
	return ParseReader(strings.NewReader(data))
}
